Add check verifying labels and annotation are not confused (#5718)

Szymongib opened the pull request:
No indication that labels and annotations were set incorrectly.

Check if labels set by user are not set as annotations and vice versa.

Unit tests and manual test.

Fixes #5718

Signed-off-by: Szymon Gibała <szymongib@gmail.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

alpeb reviewed (commented):
Looking good :+1: 
What do you think about moving the functions in `healthcheck.go` to a new file like `healthcheck_labels.go` so we keep this file size under control?
Also it'd be great to add a test case into `healthcheck_test.go` with a couple of resources violating these checks so we can ensure the output of the checks is right :wink:

dadjeibaah commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1342,6 +1342,32 @@ func (hc *HealthChecker) allCategories() []Category {
 					},
 				},
+				{
+					description: "data plane pods labels check",
```
I think we can change this description's wording to match the other checks. How about:
`data plane pod labels are configured correctly`

dadjeibaah commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1342,6 +1342,32 @@ func (hc *HealthChecker) allCategories() []Category {
+					},
+				},
+				{
+					description: "data plane services labels and annotations check",
```
Same goes for this description as well. How about:
`data plane service labels and annotations are configured correctly`

dadjeibaah commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -1898,6 +1898,145 @@ func TestValidateDataPlanePods(t *testing.T) {
 	})
 }
+
+func TestDataPlanePodLabels(t *testing.T) {
```
Thanks for adding tests 🙏 !!

dadjeibaah commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2522,3 +2557,101 @@ func containsProxy(pod corev1.Pod) bool {
+	for _, pod := range pods {
+		invalidLabels := getMisconfiguredLabels(pod.ObjectMeta)
+		if len(invalidLabels) > 0 {
+			invalid = append(invalid,
+				fmt.Sprintf("\t* pod %s/%s contains labels, which should be annotations: %s", pod.Namespace, pod.Name, strings.Join(invalidLabels, ", ")))
+		}
+	}
+
+	if len(invalid) > 0 {
+		return fmt.Errorf("Some labels on pods might be missconfigured:\n%s", strings.Join(invalid, "\n"))
```
I was playing around with the messaging around here to be more informative and let users know what exactly is misconfigured. I think this formats the error message in a way that a user can act upon it.

```suggestion
			invalid = append(invalid,
				fmt.Sprintf("\t* %s/%s\n\t\t%s", pod.Namespace, pod.Name, strings.Join(invalidLabels, "\n\t\t")))
		}
	}

	if len(invalid) > 0 {
		return fmt.Errorf("Some labels on data plane pods should be annotations:\n%s", strings.Join(invalid, "\n"))
```

The output looks like this:
```bash
‼ data plane and cli versions match
    emoji-5f947b5dbd-44fkm running git-68cebe88 but cli running dev-undefined
    see https://linkerd.io/checks/#l5d-data-plane-cli-version for hints
‼ data plane pod labels are configured correctly
    Some labels on data plane pods should be annotations:
        * emojivoto/emoji-5f947b5dbd-44fkm
                linkerd.io/inject
        * emojivoto/voting-f7f875999-phcxv
                linkerd.io/inject
                config.linkerd.io/multicluster-option
                config.linkerd.io/viz-option
    see https://linkerd.io/checks/#l5d-data-plane-pod-labels for hints
√ data plane service labels and annotations are configured correctly
```

dadjeibaah reviewed (commented):
@Szymongib, thanks for putting this PR together! This is going to be super useful. I think the change is looking great. I left a few suggestions for your consideration.

Szymongib commented:
Thanks a lot for review @alpeb and @dadjeibaah!

@alpeb a questions regarding your comment
> Also it'd be great to add a test case into `healthcheck_test.go` with a couple of resources violating these checks so we can ensure the output of the checks is right 😉

There are already some tests violating the check that expect the error https://github.com/linkerd/linkerd2/pull/5943/files#diff-932ac70a02fcd5e0e4b89b0a03f1302aeac2d162ea1b7ae7a074839d24e6dbf5R1921 do you suggest to add assertion of the actual error message?

alpeb commented:
Ah sorry I missed that, but yeah, it'd be nice to check the errors themselves as we do in the other negative tests in that file.

Szymongib commented:
I have adjusted PR to address your comments, let me know if you have any other concerns or see ways to improve.

Changes:
- Move labels and annotations checks to `healthcheck_labels.go` file
- Split services check to two to align error messages with pods check - one check for labels the other for annotations
- Change error messages according to suggestions
- Check if error message is correct in tests

dadjeibaah commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1353,6 +1353,45 @@ func (hc *HealthChecker) allCategories() []Category {
+					},
+				},
+				{
+					description: "data plane service labels configured correctly",
```
```suggestion
					description: "data plane service labels are configured correctly",
```

Szymongib replied:
Ah, thanks! I missed that!

dadjeibaah reviewed (commented):
This looks great! Thanks for making those updates. I had one minor nit but that's pretty much it

alpeb reviewed (approved):
Thanks @Szymongib , looks great! :+1:

Pothulapati reviewed (approved):
LGTM! Thanks for taking this up! 🔥 @Szymongib

Can you also update the `linkerd/website` repo with the troubleshooting docs for the newly added hint anchors? They could just mention an explanation on what labels and annotations the CLI expects for pods and services?

Szymongib commented:
Hi @Pothulapati, sure!
I have created the PR https://github.com/linkerd/website/pull/1046 let me know if that is correct place and if description is sufficient.
